libsel4vmmplatsupport/src/arch/arm/smc.c

```diff
@@ -65,8 +68,12 @@ int handle_smc(vm_vcpu_t *vcpu, uint32_t hsr)
         ZF_LOGE("Unhandled SMC: CPU service call %lu\n", fn_number);
         break;
     case SMC_CALL_SIP_SERVICE:
+#ifdef CONFIG_PLAT_ZYNQMP
+        return handle_sip(vcpu, fn_number, smc_call_is_32(id));
+#else
         ZF_LOGE("Got SiP service call %lu\n", fn_number);
         break;
+#endif
     case SMC_CALL_OEM_SERVICE:
         ZF_LOGE("Unhandled SMC: OEM service call %lu\n", fn_number);
```

Comment:
> Now that generic SMC support is added, policy should be dictated via the vm_get_smc_cap function, That means that handle_smc should be made generic to handle any kind of SMC calls, which means not limiting it to SIP services, just because that happens to be what the Zynq needs.
>
> So I think handle_smc should try to get the cap for this SMC call. If NULL callback pointer or error returned, do default handling (which is the current content of hanlde_smc). Otherwise, do the forwarding.

Yes, I think that makes sense. It's worth noting this approach gives the user a footgun to give a VM access to PSCI SMC calls to circumvent the emulation layer for PSCI, but I suppose that is fine. 

> But I would go one step further and change the vm_get_smc_cap function to a more generic vm_handle_smc callback, and if the user wants to forward the call, it can call smc_forward itself. This would give the user a chance to handle the SMC call itself any way it wants. (This would mean that the user needs to call handle_psci himself if the generic vm_handle_smc is set.
>
> Because when a VM does SMC, it tries to communicate with the secure monitor, which is actually the seL4 VMM application. The above change would make that possible. That a VM SMC call should be handled by actually the same real SMC call is in some sense coincidence.

Not sure what you mean by making `vm_get_smc_cap` more generic. All it does now (for CAmkES) is return the cap for a given function ID or a NULL if that component doesn't have that cap. This proposition sounds less generic. Or do you mean a user should be able to install a custom handler for `handle_smc`?

That could give the user more flexibility, but I am not sure I am seeing a use case for that. This only dictates how SMC calls are handled for VMs via the VMM. If your VM is making an SMC call, then you likely want to make that specific SMC call or emulate something.

Reply:
> Or do you mean a user should be able to install a custom handler for `handle_smc`?

Yes, exactly that. With `vm_get_smc_cap` all that you can do is pass some SMC calls through. With a generic smc callback the VMM can handle smc calls itself.

> That could give the user more flexibility, but I am not sure I am seeing a use case for that. This only dictates how SMC calls are handled for VMs via the VMM. If your VM is making an SMC call, then you likely want to make that specific SMC call or emulate something.

And that "emulate something" is currently missing. One example is handling suspend and resume and other PSCI calls. It's fine to provide sensible default implementations, but as it's a library, it should give full control to the user and put as little policy into the library as possible.

What a power off PSCI call should do really depends on the use case of the whole system, and libsel4vmm can't know that.

Reply:
Ok, I get what you are saying now. A user should be able to have a way to install your own emulation code at the VMM implementation instead of the VMM library. I don't really think this PR prevents that or makes it especially harder to do. I think it is adding another sane default. Another PR could add that functionality.

Reply:
I disagree. You are changing the API to add SMC support, now is the time to do it properly. The current implementation is not only unnecessarily limiting, it also makes it impossible for the user to choose between faulting the VM and returning "not implemented" error codes. And the change on your user code side is minimal, instead of returning the cap, the code just returns and calls `smc_forward()` instead.

My proposal:
- Add `vm_smc_handle_default` API function which is basically the current `handle_smc`, but exposed to users.
- Add `vm_smc_callback` pointer, which by default is set to `vm_smc_handle_default` (so the pointer is never NULL).

Then people can implement the SMC calls they want to handle, and call the default handler for the rest, or return an error if they don't want the default behaviour.

Reply:
And add a `vm_smc_set_callback()` wrapper function, like is done for other arch specific callback functions.

Reply:
> And the change on your user code side is minimal, instead of returning the cap, the code just returns and calls smc_forward() instead.

So this suggestion is to change the camkes implementation so the library (or user code) doesn't need to handle any smc capabilities itself? I suppose if that is all you ever want to do with that cap, then that is fine. 

> - Add vm_smc_handle_default API function which is basically the current handle_smc, but exposed to users.
> - Add vm_smc_callback pointer, which by default is set to vm_smc_handle_default (so the pointer is never NULL).
>Then people can implement the SMC calls they want to handle, and call the default handler for the rest, or return an error if they don't want the default behaviour.

How about this? Leave handle_smc mostly the same with the defaults that have been discussed, but before it does anything it checks to see if vm_smc_callback is set to anything. If it is, then it calls that first, and if it returns a success then handle_smc returns. Otherwise it goes through the default behaviors. 

And there will be a `vm_smc_set_callback()` function.

Reply:
> So this suggestion is to change the camkes implementation so the library (or user code) doesn't need to handle any smc capabilities itself?

No, the opposite! The user code should have full control over how to handle smc calls. I don't know what camkes does, but it shouldn't be in the way of this, it should be possible to handle (emulate) smc calls in user software build on top of camkes too.

Maybe our confusion stems from what we mean by user code: I mean end user code, users of seL4 software, while you seem to think I mean code running in user space? In my view camkes is not user code, it's an intermediate layer user code is build on top of.

In your proposal, how do you know if the function returns success? Secondly, how do you know what to do when it returns an error? Fault or return a "not implemented" error to the VM? It seems unnecessarily limiting without being simpler to implement.

Reply:
>>> And the change on your user code side is minimal, instead of returning the cap, the code just returns and calls smc_forward() instead.

>> So this suggestion is to change the camkes implementation so the library (or user code) doesn't need to handle any smc capabilities itself?

> No, the opposite! The user code should have full control over how to handle smc calls. I don't know what camkes does, but it shouldn't be in the way of this, it should be possible to handle (emulate) smc calls in user software build on top of camkes too.
>
> Maybe our confusion stems from what we mean by user code: I mean end user code, users of seL4 software, while you seem to think I mean code running in user space? In my view camkes is not user code, it's an intermediate layer user code is build on top of.

I see now, yes we were referring to different things when saying user code. I thought your first statement above was a suggested change to my CAmkES implementation of the `vm_get_smc_cap()`. I think we have an understanding on this now and the CAmkES implementation is already generic enough. 

> In your proposal, how do you know if the function returns success? 

In the usual way, with an error code return.

> Secondly, how do you know what to do when it returns an error? Fault or return a "not implemented" error to the VM? It seems unnecessarily limiting without being simpler to implement.

Yes, I suppose it is hard at that point to determine if the user attempted something and failed or just didn't implement and wants the default behavior. You could make a decision but it is a little unwieldy. I will take a stab at your suggested approach, but I think the trade-off is a user understanding on what to implement. 

If a user wants to implement an emulation or special handling of some SMC functionality, how do they know to implement `vm_smc_callback()` and that it needs to call `vm_smc_handle_default()` to retain default behavior? Updated documentation somewhere?

Reply:
> I thought your first statement above was a suggested change to my CAmkES implementation of the `vm_get_smc_cap()`. I think we have an understanding on this now and the CAmkES implementation is already generic enough.

I was suggesting to tweak your current camkes `vm_get_smc_cap()` to a version implementing `vm_smc_callback()` instead. The only change necessary in camkes code AFAIK is to change `return smc_cap;` to `return vm_smc_forward(vcpu, smc_cap);`.

> 
> > In your proposal, how do you know if the function returns success?
> 
> In the usual way, with an error code return.
> 
> > Secondly, how do you know what to do when it returns an error? Fault or return a "not implemented" error to the VM? It seems unnecessarily limiting without being simpler to implement.
> 
> Yes, I suppose it is hard at that point to determine if the user attempted something and failed or just didn't implement and wants the default behavior. You could make a decision but it is a little unwieldy. I will take a stab at your suggested approach, but I think the trade-off is a user understanding on what to implement.
> 
> If a user wants to implement an emulation or special handling of some SMC functionality, how do they know to implement `vm_smc_callback()` and that it needs to call `vm_smc_handle_default()` to retain default behavior? Updated documentation somewhere?

I would document it above the set function declaration myself. But you make a good point, there is documentation in the doc directory, so it could also be added under the ARM specific documentation.

Assuming the camkes code is automatically generated and it will call the set function, it's probably good to give camkes' function a well defined name and document that in the camkes docs, so people that want to override default behaviour while using camkes generated pass-through code can call that instead of `vm_smc_handle_default()`. I don't have camkes experience myself, so I don't know whether it exposes any lower level stuff or if it's all hidden. If the latter, I understand why my proposal didn't make much sense to you, as it's pointless for camkes, until it adds explicit support for smc emulation.

Reply:
Ok, I implemented your proposed updates. I documented the register functions.